Tactics/CSE.lean

```diff
@@ -154,46 +193,61 @@ private def Array.replicate (n : Nat) (v : α) : Array α := List.replicate n v
 def ExprData.isProfitable? (data : ExprData) : CSEM Bool :=
   return data.size > 1 && data.occs >= (← getConfig).minOccsToCSE
 
+/-- Check if an expression is a nat literal, or a `OfNat.ofNat` of a nat literal.
+This lets us avoid CSEing over nat literals, which are already in canonical form.
+-/
+def CSEM.isNatLit (e : Expr) : Bool :=
+  if e.isRawNatLit then true
+  else
+  match_expr e with
+  | OfNat.ofNat _α x _inst  => x.isRawNatLit -- @OfNat.ofNat Nat 1 (instOfNatNat 1)
+  | _ => false
```

Comment:
```lean
/-- Check if an expression is a nat literal, or a `OfNat.ofNat` of a nat literal.
This lets us avoid CSEing over nat literals, which are already in canonical form.
-/
def CSEM.isNatLit (e : Expr) : Bool :=
  if e.isRawNatLit then true
  else
  match_expr e with
  | OfNat.ofNat _α x _inst  => x.isRawNatLit -- @OfNat.ofNat Nat 1 (instOfNatNat 1)
  | _ => false
```

I had this code before, and then lost it. I brought it back to make sure that we don't spend ages CSEing stuff like natural number literals.